homeassistant/components/mikrotik/__init__.py

```diff
@@ -1 +1,274 @@
+                 user, password, login_method, encoding):
+        """Initialize the Mikrotik Client."""
+        self.hass = hass
+        self._host = host
+        self._ssl = ssl
+        self._port = port
+        self._user = user
+        self._password = password
+        self._login_method = login_method
+        self._encoding = encoding
+        self._host_name = ''
+        self._client = None
+        self._connecting = False
+        self._connected = False
+
+    def connect_to_device(self):
+        """Connect to Mikrotik method."""
+        if self._connecting:
+            return
+        self._connecting = True
+        _LOGGER.debug("[%s] Connecting to Mikrotik device.", self._host)
+
+        kwargs = {
+            'encoding': self._encoding,
+            'login_methods': self._login_method,
+            'port': self._port
+        }
+
+        if self._ssl:
+            ssl_context = ssl.create_default_context()
+            ssl_context.check_hostname = False
+            ssl_context.verify_mode = ssl.CERT_NONE
+            kwargs['ssl_wrapper'] = ssl_context.wrap_socket
+
+        try:
+            self._client = librouteros.connect(
+                self._host, self._user, self._password, **kwargs)
+        except (librouteros.exceptions.TrapError,
+                librouteros.exceptions.MultiTrapError,
+                librouteros.exceptions.ConnectionError) as api_error:
+            _LOGGER.error(
+                "Mikrotik error for device %s. "
+                "Connection error: %s", self._host, api_error)
+            self._connecting = False
+            self._connected = False
+            self._client = None
+            return False
+
+        cmd = MIKROTIK_SERVICES[IDENTITY]
+        self._host_name = (self._client(cmd=cmd))[0]['name']
+        if not self._host_name:
+            _LOGGER.error("Mikrotik failed to connect to %s.",
+                          self._host)
+            return False
+        _LOGGER.info("Mikrotik Connected to %s (%s).",
+                     self._host_name, self._host)
+        self._connecting = False
+        self._connected = True
+        return True
+
+    def get_hostname(self):
+        """Return device host name."""
+        if not self._connected:
+            self.connect_to_device()
+        return self._host_name
+
+    def connected(self):
+        """Return connected boolean."""
+        return self._connected
+
+    async def update_info(self):
+        """Update info from Mikrotik API."""
+        _LOGGER.debug("[%s] Updating Mikrotik info.", self._host)
+        if not self._connected:
+            self.connect_to_device()
+        data = self.get_api('/system/routerboard/getall')
+        if data is None:
+            _LOGGER.error(
+                "Mikrotik device %s is not connected.",
+                self._host)
+            self._connected = False
+            return
+        self.hass.data[DOMAIN][self._host]['info'] = data[0]
```

Comment:
Why save all these things on hass.data instead of on this api instance?

I think we should just save the api instances, one per client, in hass.data. All information that can be organized per client or host, should be stored on the api instance.

Reply:
That does make sense. I will update the api instance to store the data instead of using hass.data.